### DISCUSSION TITLE
Version bumping feature

### DISCUSSION THREAD

**Discussion by davidkron** [jreleaser/jreleaser#595] [Ideas]
Hi! First of all, really cool project. I haven't tried it yet, but definitely will someday.

My first thought on this project was that this really could become to the Java ecosystem, what semantic-release managed to do in the Node world, but more flexible and less opinionated. But for this (at least for me), it would definitely require some version bumping feature to be included (definitely not as a default, but to be configurable when desired).

I do know that there are other libraries or build plugins available for this task, but I have evaluated some of the very popular ones and I found most to be either lacking in features or flexibility or they were tightly coupled to a single tool (Maven, Gradle). This is especially cumbersome if a company (like mine) is not fixated on a single build tool and wants to standardize a consistent release process independent of the tool. I like that JReleaser is implemented as a core library with adapters for Maven, Gradle and even standalone usage (CLI, Docker). This is a great selling point.

Another argument for me to have a version bump feature in JReleaser is that it already supports changelog generation/evaluation using conventional commits. Some releasing workflows today are automatically bumping versions (usually with semver as a target) based on conventional commits. So a version bumper will determine which number part to bump based on the categorized/evaluated changelog. If both implementations are coming from the same library it will definitely yield better results than two separate libraries (each with their own flaws/bugs = inconsistent results).

Third, a deficiency I found in almost all the build plugins was that they were fixated on a single opinionated version schema (semver, calver, 0-ver, chronver, ...). But different companies and projects exist and they have their reasons and opinions for using one or the other. So "pluggable" version schemas could be a very good selling point which imho would set this library immediately apart from others.

What are your thoughts?

**Comment by aalmiray** [comment]
Hi @davidkron, hope you give JReleaser a try in the near future.

We currently support 5 versioning schemes as shown at https://jreleaser.org/guide/latest/configuration/project.html#_version, mainly

 1. semver
 2. calver
 3. java module
 4. java runtime
 5. custom

We could calculate an increment for the first 4 but can't make it for the `custom` one. Now, the rules for auto-incrementing a verison number vary not only based on the version schema but also on a particular wish, say sometimes one would like to bump from 1.2.3 to 1.3.0 even though the algorithm tells you it should be 1.2.4 for example.

Another thing to ponder is: does updating the project version constitute a job of the release tool, the build tool, or something else? Certainly it's part of the release process but given that as of today JReleaser does not publish JARs and POMs to Maven Central (that is still a job of the build tool of choice) one wonders where are the boundaries between build tool and release tool. I do agree that publishing JARs and POMs are part of the release process.

**Comment by yschimke** [comment]
I think this tweet https://twitter.com/yschimke/status/1475193457635385348 is really a request for this feature.

I mostly just want to bump minor version and release for my github project.  My ideal work flow would be

1) Make changes and commit to master
2) ./gradlew jreleaseTriggerRelease (tags the current commit, pushes to github).
3) Have a github action I've setup run on each tag matching a pattern and complete the ./gradlew jreleaserFullRelease

What doesn't work well currently

- nemerosa defaults want releases off a branch "release/3.12" and tags like "v3.12.0"
- I'd like to initiate a release just by creating a tag "v3.12.0" - this requires a releaseParser and still ends up with a version 3.12.0
- Even after dealing with this I'm getting the CI jreleaserFullRelease step fail with `Version v3.12 does not follow the semver spec`

I looked at jreleaser and it works off workflow_dispatch build type and `github.event.inputs.version` but still has a lot of handling of versions.  

It would be nice if jreleaser didn't require config of versioning plugins and/or github action steps to get things working.

For reference

Github workflow: https://github.com/yschimke/okurl/blob/main/.github/workflows/release.yml
Nemerosa releaseParser: https://github.com/yschimke/okurl/blob/main/build.gradle.kts#L18-L28

**Reply by aalmiray** [reply]
I see a few things that need unpacking here. First, what's the state of the version number at step 1? I'd assume it's some sort of snapshot version (let's say it has the -SNAPSHOT tag). What do you expect to happen at step 2? I suppose the version number becomes releasable i.e, looses -SNAPSHOT, a tag matching the version number (or maybe v${version}) is created then pushed to the remote Git repository, where a GitHub action with a tag trigger reacts and continues the process, right?

If so, there are already several plugins in the Gradle ecosystem that can do just that, why would JReleaser be another one? Say, JBang uses https://github.com/toolebox-io/gradle-git-versioner. JReleaser itself does not use any auto-versioning scheme because personally I like to micro-manage that particular aspect. If I didn't I'd pick a Gradle plugin that performs that task.

There's no limitation from JReleaser's POV to initiate a release. You can do it manually or using a GitHub Action trigger. You can tag the release on your own, using another plugin, or let JReleaser do it for you. If the tag is created externally you can instruct JReleaser to skip creating the tag (as it'll create the tag by default). This means you're in full control on how and when the tag is created.

If JReleaser complains with `Version v3.12 does not follow the semver spec` it means that the project version is set to `v3.12` when it should be `3.12`. This may be a problem caused by the tag parsing on your build.

Your project has lots of tags, but following the defaults only these ones will be taken into account:

```
v3.10.0
v3.11.0
v3.11.0.0
v3.12
v3.13
v3.14
v3.7
v3.8.0
v3.9.0
```

With `v3.11.0.0` looking a bit iffy as the last `0` would be treated as a tag.

GitHub actions are not needed for JReleaser to do its job. Actually that is one of its core strengths, as it can function anywhere as long as there's a Git repository, with or without Java installed, locally or remote. In the case of building Graal Native Images you  do need to run the build on different platforms if you want to create separate binaries, in this case GitHub Actions are ideal as configuring a build matrix for the 3 major OSes is a snap.

**Reply by yschimke** [reply]
> why would JReleaser be another one?

Understood, no problem with your decision.  My reasoning was that you need a version system that works across github, gradle/nemerosa and jreleaser.  A default one in jreleaser that fit that bill would be vastly simpler *for me* :)

**Reply by yschimke** [reply]
> This may be a problem caused by the tag parsing on your build.

I think so, a test build hardcoded to "3.14.0" got further and failed on cross repo writes for brew.  I'll try to debug how it's failing like that.  Thanks.

**Reply by yschimke** [reply]
Nothing I can't do myself, but even just having support for jreleaser triggering the release, seems like completing the end to end story.  But understood if you don't feel it should be a core part of jreleaser.  The amount of bespoke version config in the jreleaser builds I've seen (custom tasks writing to version files and so on), does suggest it's an unsolved problem still.

**Reply by aalmiray** [reply]
> Understood, no problem with your decision. My reasoning was that you need a version system that works across github, gradle/nemerosa and jreleaser. A default one in jreleaser that fit that bill would be vastly simpler _for me_ :)

Right, I understand. However nemerosa's versioning understands semver while JReleaser supports more versioning schemes than semver, some of which are completely incompatible with semver. If JReleaser were to offer such feature it would have to do so with all versioning schemes, which is of course a bit of extra work.

> Nothing I can't do myself, but even just having support for jreleaser triggering the release, seems like completing the end to end story. But understood if you don't feel it should be a core part of jreleaser. The amount of bespoke version config in the jreleaser builds I've seen (custom tasks writing to version files and so on), does suggest it's an unsolved problem still.

Understood. for the previous reason stated of multiple versioning schemes there's no support from JReleaser to set a version number based on commit information, or moving from snapshot to release version. Note that JReleaser supports more than one mode to determine if a version is snapshot or not, using matching pattern under your control (`*-SNAPSHOT` being the default pattern).

My suggestion would be to explore the existing Gradle release plugins to see if one of them provides the behavior you seek. The one I linked (used by JBang) appears to be pretty good. If you were to rely on it then your release workflow would look like:

1. commit, push, commit, push, til you're ready.
2. create a commit whose message has a tag supported by the Git versioner plugin i.e, `[minor]`, or `[patch]`.
3. this updates the project version to the next number according to the settings, also creates a tag, pushes commit and tag to remote repo.
4. you have a GH action that triggers on tags. This action is responsible for invoking JReleaser.

With this there's no need to have bespoke version handling code as JReleaser does in its own build.

**Reply by aalmiray** [reply]
One more thing, you can try a release locally in dryrun mode to see how it would play out, no need to push commits to GitHub actions, unless your intention is to excercise a GH action.

**Reply by yschimke** [reply]
I tried and failed to work out how to configure that on command line.

Just saw

```
dryrun = (rootProject.findProperty('jreleaser.dryrun') ?: false).toBoolean()
```

I was expecting to see it here

```
 ./gradlew help --task=jreleaserFullRelease
...
Options
     --select-current-platform     Activates paths matching the current platform (OPTIONAL).

     --select-platform     Activates paths matching the given platform (OPTIONAL).
```

**Reply by aalmiray** [reply]
At the moment the dryrun setting is a property on the `jreleaser` extension, thus must be set explicitly with custom property parsing as shown in the documentation (1st snippet you pasted).

I've been thinking if that setting should be available as a command flag as well. If set as a command flag would have precedence over any value set in the extension.

**Comment by yschimke** [comment]
This particular aspect is the bane of my usage of jreleaser.  It's not jreleasers fault, but having tools opinionated on release versions like jreleaser, jitpack, _XXX_ gradle version plugin, github, makes it awkward to get a working solution.

Maybe there is another way forward, like a blessed example showing github, gradle that is known to work well.

If I put up an toy project somewhere, but publishing to these tools, can I get some advice on how to configure it?

**Reply by aalmiray** [reply]
As you may be aware by now, JReleaser took inspiration from GoReleaser which has this to say regarding versioning -> https://goreleaser.com/limitations/semver/

```
Your tag should be a valid semantic version. If it is not, GoReleaser will error.
```

We believe versioning should be more flexible and thus we support multiple versioning schemes. The downside of this is that JReleaser does not handle version auto-increment for you. It's flexible enough to let you pick your own scheme but you must figure out how to update the version when needed. More over, JReleaser may be invoked in multiple ways, not just as a Gradle plugin, thus the behavior provided must work for all possible invocations.

**Comment by gotson** [comment]
I agree with @davidkron, if we could get something like what semantic release does, that would be awesome.

I use semantic release, and the bumping of version is excellent. Using a plugin I can replace the version in my gradle.prooerties, and commit it back to master.

**Reply by aalmiray** [reply]
Isn't the case that there's already plugins for your build tool of choice that take care of version bumping according to Semantic Versioning rules? If so it's a mater of choosing the one that works for you. Unless thew wish is to have JReleaser be the provider of all behavior, which at this point I'm still on the fence for the same reason it does not publish JARs and POMs to a Maven compatible remote repository (such as Maven Central) -> build tools already know how to do this.

**Reply by gotson** [reply]
I think the point made above is very valid, if jreleaser handles the Changelog, it should handle the version bump.

The publishing to maven is probably the only thing done by the build tool, and it's only valid for libraries. I'm developing an app so that's not even something I do.

I'd like to switch to jreleaser because of the possibilities around publishing to various repos, and jpackage and native images.

**Reply by aalmiray** [reply]
Why would that be the case? The changelog is a list of items from tag A to tag B or from latest commit to latest tag. The typical use case is the later, however you can request calculating the changelog for 2 tags somewhere within the tag list, not necessarily latest.

What does calculating the changelog have to do with bumping the version?

**Reply by gotson** [reply]
Sorry, I was thinking of conventional commits. Since you already handle those, it makes sense to be able to compute a version bump according to those rules.

**Reply by gotson** [reply]
To complement about the tool I use, semantic release is geared toward javascript, even though it can handle more.

It has limitations though, for instance most of the behaviors I need must be handled via custom logic or custom plugins, like building the docker image for instance.

So the only stock features I get are :
- compute version bump
- generate Changelog
- create github release (with Changelog content)
- commit back some files, like the changed gradle.properties and the Changelog.md, to master

Looking at the list, jreleaser already handles Changelog, github release, and I see there is a planned change for handling an external Changelog.md.

So what's left is just computing a version bump, and committing back.

Im just saying that a single tool around releases would really help, and I haven't crossed any good gradle plugins for semantic versioning yet.

And as someone else made the point in this discussion I think, the fact they jreleaser is a library with multiple bindings really shines. You're not bound to maven or gradle, and you have coherent behaviors whatever the tool used.

**Reply by aalmiray** [reply]
Got it. I missed the part where CC refers to semver, which is https://www.conventionalcommits.org/en/v1.0.0/#how-does-this-relate-to-semver

```
fix type commits should be translated to PATCH releases. feat type commits should be translated to MINOR releases. Commits with BREAKING CHANGE in the commits, regardless of type, should be translated to MAJOR releases.
```

Let's assume JReleaser can calculate an appropriate version bump. The next step is, what to do with it and when? Right now JReleaser expects the project version as an input as there may or may not be an existing tag for the current version. This is one of the aspects where JReleaser differs from GoReleaser as the latter expects a tag to exist before performing a release. Why? Because how else would you set the version for a Go project?

If the project version is kept as an input then JReleaser would only be able to calculate the next version after release. My gut feeling tells me you'd want JReleaser to calculate the current version, isn't that right?

Say it's OK to keep supplying the current version as an explicit value, JReleaser calculates the next version. Where should it go? What should JReleaser do next? If the project uses the Maven plugin then updating the POM(s) and committing is typically done via the `version` or `release` plugins. For Gradle things start to get more complicated as the version may be found inside `gradle.properties` or `build.gradle[.kts]` using the standard `version` property, or it could be done in a totally different way. For example JReleaser itself uses an external file named `VERSION`. For Ant you'd have to come up with your own conventions. And if you're using JReleaser as a CLI there's no tie to a specific build tool.

So what JReleaser can do after calculating the next version is to save it alongside all other calculated values in `[out|target|build]/jreleaser/output.properties` where another command can read the value and update the project specific resource.

**Reply by gotson** [reply]
> My gut feeling tells me you'd want JReleaser to calculate the current version, isn't that right?

Not necessarily, not in my case for instance. I think the behaviour could be optional, but also configurable.

It could be given the previous version from the last tag, or a given tag, and compute the current version by looking at the commits between the tag provided and `HEAD`.



> Say it's OK to keep supplying the current version as an explicit value, JReleaser calculates the next version. Where should it go? What should JReleaser do next?

Again here, optional, and configurable. In my particular case i output it inside `gradle.properties` and replace the previous version. This is needed because the current version is part of the jar built by gradle, and is used at runtime (you can query an API endpoint to get the version number).

Other workflows may require something else. I know Maven have those version plugins which does a lot of things, but i don't like those at all. Some people do.

**Comment by gotson** [comment]
@aalmiray now that we have Conventional Commit parsing, do you think we could get closer to a version bumping feature in JReleaser?

In order to automate version bumping using Maven/Gradle for conventional commits, i'm currently using different tools:
- some tools to perform the commit history analysis and output the next version, currently i use a mix of https://github.com/caarlos0/svu and https://github.com/semantic-release/semantic-release
- for Maven, i use the Maven version plugin to replace the version in `pom.xml`
- for Gradle, i don't have a good way to update the version in `gradle.properties` (I even asked on [SO](https://stackoverflow.com/questions/74441421/how-to-update-property-in-gradle-properties-file) without success). semantic-release has [a replace plugin](https://github.com/google/semantic-release-replace-plugin) that helps in that regards, but it's coupled to the tool.

**Reply by aalmiray** [reply]
The problem we face at the moment is that the project version is an expect input. Version bump can only happen after processing the changelog, which only occurs as a dependency for most but not all steps. For example, assemble, download, deploy, and config do not depend on changelog.

**Reply by gotson** [reply]
Could it be a jreleaser capability, but that would work separately from the rest? It could be taken out of the changelog task, and ran before.

In grade it would be probably easier, with the tasks dependencies, not sure about maven, ant or cli though.

I wouldn't mind having the capability separated from the whole jreleaser full release. It would at least bring another important part of releasing under a single tool, and would also have consistency with changelog generation.

**Reply by aalmiray** [reply]
This certainly requires some thought. The project version is used to determine the tag pattern to find out the previous tag. If there’s no project version as input then JReleaser would have to:
- enforce SEMVER as version pattern, otherwise fail.
- expect a tag pattern as input, otherwise fail. Why is a tag pattern required? Some use . instead of - to separate the tag from the number (looking at you Quarkus).

Given these conditions JReleaser would know the next version must be computed.

**Reply by aalmiray** [reply]
Looking at caarlos0/svu it seems we have what we need to replicate the behavior (thanks for the PRs!) so that it could be extracted into its own space. Maybe even invoked as a base step before all others, also able to invoked separate from all steps (if you only care to know the version number).

As a separate step it could also have a setting to commit the version, probably as part of the `release` block. At the very least it would need a property to mark it as active/enabled (make a commit or not) and the commit message template.
